Nullexception in `OAuth2MessageFactory` for cached token

Ive experienced a few times now that when downloading a collection it does not download all of the mods associated with the collection. 

Has anyone else experienced this?

I saw this exception as well, might be a concurrency issue.

```
00:16:17.162 [ERROR] (NexusMods.SingleProcess.CommandLineConfigurator) Exception while running handler NexusMods.CLI.Types.IpcHandlers.NxmIpcProtocolHandler for `nxm://cyberpunk2077/collections/iszwwe/revisions/469`|System.InvalidOperationException: Nullable object must have a value.
   at NexusMods.Networking.NexusWebApi.Auth.OAuth2MessageFactory.GetOrRefreshToken(CancellationToken cancellationToken) in /_/src/Networking/NexusMods.Networking.NexusWebApi/Auth/OAuth2MessageFactory.cs:line 49
   at NexusMods.Networking.NexusWebApi.Auth.OAuth2MessageFactory.Create(HttpMethod method, Uri uri) in /_/src/Networking/NexusMods.Networking.NexusWebApi/Auth/OAuth2MessageFactory.cs:line 73
   at NexusMods.Networking.NexusWebApi.NexusApiClient.ModInfoAsync(String domain, ModId modId, CancellationToken token) in /_/src/Networking/NexusMods.Networking.NexusWebApi/NexusApiClient.cs:line 186
   at NexusMods.Networking.NexusWebApi.NexusModsLibrary.GetOrAddModPage(ModId modId, GameDomain gameDomain, CancellationToken cancellationToken) in /_/src/Networking/NexusMods.Networking.NexusWebApi/NexusModsLibrary.cs:line 46
   at NexusMods.Networking.NexusWebApi.NexusModsLibrary.CreateDownloadJob(AbsolutePath destination, GameDomain gameDomain, ModId modId, FileId fileId, Optional`1 nxmData, CancellationToken cancellationToken) in /_/src/Networking/NexusMods.Networking.NexusWebApi/NexusModsLibrary.cs:line 216
   at NexusMods.Collections.InstallCollectionJob.EnsureNexusModDownloaded(Mod mod) in /_/src/NexusMods.Collections/InstallCollectionJob.cs:line 218
   at NexusMods.Collections.InstallCollectionJob.EnsureNexusModDownloaded(Mod mod) in /_/src/NexusMods.Collections/InstallCollectionJob.cs:line 222
   at NexusMods.Collections.InstallCollectionJob.EnsureDownloaded(Mod mod) in /_/src/NexusMods.Collections/InstallCollectionJob.cs:line 199
   at NexusMods.Collections.InstallCollectionJob.<>c__DisplayClass33_0.<<StartAsync>b__1>d.MoveNext() in /_/src/NexusMods.Collections/InstallCollectionJob.cs:line 80
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.<>c__53`1.<<ForEachAsync>b__53_0>d.MoveNext()
--- End of stack trace from previous location ---
   at NexusMods.Collections.InstallCollectionJob.StartAsync(IJobContext`1 context) in /_/src/NexusMods.Collections/InstallCollectionJob.cs:line 80
   at NexusMods.Jobs.JobContext`2.Start() in /_/src/NexusMods.Jobs/JobContext.cs:line 43
   at NexusMods.CLI.Types.IpcHandlers.NxmIpcProtocolHandler.HandleCollectionUrl(NXMCollectionUrl collectionUrl) in /_/src/NexusMods.App.Cli/Types/IpcHandlers/NxmIpcProtocolHandler.cs:line 120
   at NexusMods.CLI.Types.IpcHandlers.NxmIpcProtocolHandler.HandleCollectionUrl(NXMCollectionUrl collectionUrl) in /_/src/NexusMods.App.Cli/Types/IpcHandlers/NxmIpcProtocolHandler.cs:line 120
   at NexusMods.CLI.Types.IpcHandlers.NxmIpcProtocolHandler.Handle(String url, CancellationToken cancel) in /_/src/NexusMods.App.Cli/Types/IpcHandlers/NxmIpcProtocolHandler.cs:line 82
   at NexusMods.SingleProcess.CommandLineConfigurator.RunLink(String[] args, CancellationToken cancellationToken) in /_/src/NexusMods.SingleProcess/CommandLineConfigurator.cs:line 119

```

--- a/src/Networking/NexusMods.Networking.NexusWebApi/Auth/JWTToken.cs
+++ b/src/Networking/NexusMods.Networking.NexusWebApi/Auth/JWTToken.cs
@@ -1,10 +1,8 @@
-using System.Diagnostics.CodeAnalysis;
+using DynamicData.Kernel;
 using NexusMods.Abstractions.NexusWebApi.DTOs.OAuth;
 using NexusMods.MnemonicDB.Abstractions;
 using NexusMods.MnemonicDB.Abstractions.Attributes;
 using NexusMods.MnemonicDB.Abstractions.Models;
-using NexusMods.MnemonicDB.Abstractions.TxFunctions;
-using File = NexusMods.Abstractions.Loadouts.Files.File;
 
 namespace NexusMods.Networking.NexusWebApi.Auth;
 
@@ -29,49 +27,50 @@ public partial class JWTToken : IModelDefinition
     /// The date at which the token expires
     /// </summary>
     public static readonly TimestampAttribute ExpiresAt = new(Namespace, nameof(ExpiresAt));
-    
-    
+
+    private static Optional<EntityId> GetEntityId(IDb db)
+    {
+        var datoms = db.Datoms(PrimaryAttribute);
+        return datoms.Count == 0 ? Optional<EntityId>.None : datoms[0].E;
+    }
+
     /// <summary>
     /// Try to find the JWT Token in the database.
     /// </summary>
     public static bool TryFind(IDb db, out ReadOnly token)
     {
-        var found = All(db).FirstOrDefault();
-        if (found.IsValid())
+        var entityId = GetEntityId(db);
+        if (!entityId.HasValue)
         {
-            token = found;
-            return true;
+            token = default(ReadOnly);
+            return false;
         }
-        
-        token = default(ReadOnly);
-        return false;
+
+        token = Load(db, entityId.Value);
+        return token.IsValid();
     }
-    
-    
+
     /// <summary>
     /// Creates a new JWT Token model from a <see cref="JwtTokenReply"/>. And reuses the existing
     /// database id if it exists, as this data is a singleton.
     /// </summary>
-    public static EntityId? Create(IDb db, ITransaction tx, JwtTokenReply reply)
+    public static Optional<EntityId> Create(IDb db, ITransaction tx, JwtTokenReply reply)
     {
-        if (reply.AccessToken is null || reply.RefreshToken is null) 
-            return null;
-            
-        var existingId = db.Datoms(JWTToken.AccessToken).FirstOrDefault().E;
-        if (existingId == EntityId.From(0))
-            existingId = tx.TempId();
-        
-        tx.Add(existingId, JWTToken.AccessToken, reply.AccessToken);
-        tx.Add(existingId, JWTToken.RefreshToken, reply.RefreshToken);
-        tx.Add(existingId, JWTToken.ExpiresAt, DateTimeOffset.FromUnixTimeSeconds(reply.CreatedAt).DateTime + TimeSpan.FromSeconds(reply.ExpiresIn));
+        if (reply.AccessToken is null || reply.RefreshToken is null) return Optional<EntityId>.None;
 
-        return existingId;
+        var existingId = GetEntityId(db);
+        var entityId = existingId.HasValue ? existingId.Value : tx.TempId();
+
+        tx.Add(entityId, AccessToken, reply.AccessToken);
+        tx.Add(entityId, RefreshToken, reply.RefreshToken);
+        tx.Add(entityId, ExpiresAt, DateTimeOffset.FromUnixTimeSeconds(reply.CreatedAt).DateTime + TimeSpan.FromSeconds(reply.ExpiresIn));
+
+        return entityId;
     }
     
     /// <summary>
     /// Model for the JWT Token
     /// </summary>
-    /// <param name="tx"></param>
     public partial struct ReadOnly
     {
         /// <summary>

--- a/src/Networking/NexusMods.Networking.NexusWebApi/Auth/OAuth2MessageFactory.cs
+++ b/src/Networking/NexusMods.Networking.NexusWebApi/Auth/OAuth2MessageFactory.cs
@@ -1,14 +1,8 @@
-using System.Reactive.Linq;
-using DynamicData.Binding;
 using Microsoft.Extensions.Logging;
-using NexusMods.Abstractions.MnemonicDB.Attributes;
 using NexusMods.Abstractions.NexusWebApi;
 using NexusMods.Abstractions.NexusWebApi.DTOs.OAuth;
 using NexusMods.Abstractions.NexusWebApi.Types;
-using NexusMods.Abstractions.Serialization;
-using NexusMods.Extensions.BCL;
 using NexusMods.MnemonicDB.Abstractions;
-using NexusMods.MnemonicDB.Abstractions.Query;
 
 namespace NexusMods.Networking.NexusWebApi.Auth;
 
@@ -31,31 +25,23 @@ public class OAuth2MessageFactory : IAuthenticatingMessageFactory
         _conn = conn;
         _auth = auth;
         _logger = logger;
-        
-        _conn.ObserveDatoms(SliceDescriptor.Create(JWTToken.AccessToken, _conn.Registry))
-            .Subscribe(_ => _cachedTokenEntity = null);
     }
 
-    private JWTToken.ReadOnly? _cachedTokenEntity;
     private readonly IConnection _conn;
 
     private async ValueTask<string?> GetOrRefreshToken(CancellationToken cancellationToken)
     {
-        if (!JWTToken.TryFind(_conn.Db, out var token))
-            return null;
-        
-        _cachedTokenEntity = token;
-        if (!token.HasExpired) 
-            return _cachedTokenEntity!.Value.AccessToken;
+        if (!JWTToken.TryFind(_conn.Db, out var token)) return null;
+        if (!token.HasExpired) return token.AccessToken;
 
         _logger.LogDebug("Refreshing expired OAuth token");
 
         var newToken = await _auth.RefreshToken(token.RefreshToken, cancellationToken);
         var db = _conn.Db;
         using var tx = _conn.BeginTransaction();
-        
+
         var newTokenEntity = JWTToken.Create(db, tx, newToken!);
-        if (newTokenEntity is null)
+        if (!newTokenEntity.HasValue)
         {
             _logger.LogError("Invalid new token in OAuth2MessageFactory");
             return null;
@@ -63,8 +49,8 @@ public class OAuth2MessageFactory : IAuthenticatingMessageFactory
 
         var result = await tx.Commit();
 
-        _cachedTokenEntity = JWTToken.Load(result.Db, result[newTokenEntity.Value]);
-        return _cachedTokenEntity!.Value.AccessToken;
+        token = JWTToken.Load(result.Db, result[newTokenEntity.Value]);
+        return token.AccessToken;
     }
 
     /// <inheritdoc/>

--- a/src/Networking/NexusMods.Networking.NexusWebApi/LoginManager.cs
+++ b/src/Networking/NexusMods.Networking.NexusWebApi/LoginManager.cs
@@ -135,7 +135,7 @@ public sealed class LoginManager : IDisposable, ILoginManager
         using var tx = _conn.BeginTransaction();
 
         var newTokenEntity = JWTToken.Create(_conn.Db, tx, jwtToken);
-        if (newTokenEntity is null)
+        if (!newTokenEntity.HasValue)
         {
             _logger.LogError("Invalid new token data");
             return;

